File: Java/Spring-Boot/exception-handler-junit/exception-handler-junit/src/main/java/jp/glory/sample/web/handler/WebExceptionHandler.java
package jp.glory.sample.web.handler;

import org.springframework.beans.TypeMismatchException;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

import jp.glory.sample.exceptioin.OriginalException;
import jp.glory.sample.web.response.ErrorResponse;

@ControllerAdvice
public class WebExceptionHandler extends ResponseEntityExceptionHandler {


    @ExceptionHandler(OriginalException.class)
    public ResponseEntity<ErrorResponse> handleOriginalException(OriginalException ex) {

        ErrorResponse response = new ErrorResponse();
        response.setMessage("Throw OriginalException!!");

        return new ResponseEntity<ErrorResponse>(response, HttpStatus.BAD_REQUEST);
    }

    @Override
    protected ResponseEntity<Object> handleTypeMismatch(TypeMismatchException ex, HttpHeaders headers,
            HttpStatus status, WebRequest request) {

        ErrorResponse response = new ErrorResponse();
        response.setMessage("Throw TypeMismatchException!!");

        return new ResponseEntity<Object>(response, HttpStatus.BAD_REQUEST);
    }
}

File: Measurement/practice-measure-architecture/src/main/kotlin/jp/glory/practice/measure/architecture/base/adaptor/web/WebExceptionHandler.kt
package jp.glory.practice.measure.architecture.base.adaptor.web

import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Component
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler

@Component
class WebExceptionHandler : ResponseEntityExceptionHandler() {
    private val logger = LoggerFactory.getLogger(this::class.java)

    @ExceptionHandler(WebException::class)
    fun webException(
        ex: WebException,
    ): ResponseEntity<ProblemDetail> =
        when (ex.error) {
            is WebUnknownError -> handleWebUnknownError(ex.error)
            is WebNotFoundError -> handleNotFoundError(ex.error)
            is WebValidationError -> handleValidationError(ex.error)
        }
            .let {
                ResponseEntity.status(it.status).body(it)
            }

    private fun handleWebUnknownError(
        error: WebUnknownError
    ): ProblemDetail {
        logger.error(error.message, error.cause)

        return ProblemDetail.forStatus(HttpStatus.INTERNAL_SERVER_ERROR)
            .apply {
                detail = "Unknown error is occurred."
            }
    }

    private fun handleNotFoundError(
        error: WebNotFoundError
    ): ProblemDetail =
        ProblemDetail.forStatus(HttpStatus.NOT_FOUND)
            .apply {
                detail = "${error.resourceName} is not found."
                setProperty("id", error.idValue)
                setProperty("resource", error.resourceName)
            }

    private fun handleValidationError(
        error: WebValidationError
    ): ProblemDetail {
        val problemDetail = ProblemDetail.forStatus(HttpStatus.BAD_REQUEST)
            .apply {
                detail = "Invalid input."
            }

        error.details.forEachIndexed { index, detail ->
            problemDetail.setProperty("id-${index}", index)
            problemDetail.setProperty("message-${index}", detail.createMessage())
        }

        return problemDetail
    }
}